Stock sync

DanielePalombo opened the pull request:
When a shipwire's webhook calls 'stock_controller/create' the action
updates `count_on_hand` in `stock_item` and adds a `stock_movement` to
keep track movements.

DanielePalombo committed:
Add shipwire webhook controller

DanielePalombo committed:
Add shipwire stock_sync controller

DanielePalombo committed:
Missing require in Engine

DanielePalombo committed:
Add Spree:SolidusShipwirePreference

Preferences can add in solidus standard way

DanielePalombo committed:
Upgrade to rspec 3

DanielePalombo committed:
Add Spree::ShipwireWebhookController

Spree::ShipwireWebhookController can be inherited by other controller that receive webhook.
The action in this controller can be executed only if the request is valid

DanielePalombo committed:
Add helper to stub signature

In test the signature can be stubbed using `stub_signature!`

DanielePalombo committed:
Add shipwire_id to Spree::StockLocation

DanielePalombo committed:
Add Spree::ShipwireWebhook::StockController

When a shipwire's webhook calls 'stock_controller/create' the action
updates `count_on_hand` in `stock_item` and adds a `stock_movement` to
keep track movements.

kennyadsl commented on config/routes.rb:
```diff
@@ -1,3 +1,9 @@
+    resources :stock, only: :create
+
+    # match all HEAD from shipwire for webhook subscription
+    match '*path' => '/spree/shipwire_webhook#subscribe', :via => :head
```
`via: :head`

kennyadsl commented on app/controllers/spree/shipwire_webhook_controller.rb:
```diff
@@ -0,0 +1,34 @@
+class Spree::ShipwireWebhookController < ActionController::Base
+  respond_to :json
+
+  before_action :validate_key, except: :subscribe
```
this is done except `subscribe` but  `subscribe` seems to be the only action here. Am I missing something?

DanielePalombo replied:
This controller can be extended by others controller. All actions need to be validated before execute

kennyadsl replied:
shouldn't be the others controllers to take care of not running that callback?

mtylty replied:
I think this looks good. It's a basic check it shouldn't hurt other code and it's covered with specs.

kennyadsl commented on lib/solidus_shipwire/engine.rb:
```diff
@@ -10,6 +11,14 @@ class Engine < Rails::Engine
     end
 
+    initializer 'solidus_shipwire.environment', before: 'spree.environment' do
+      Spree::SolidusShipwireConfig = Spree::SolidusShipwireConfiguration.new
```
why not use `Spree::ShipwireConfig`? If one day`Spree` namespace will be converted into `Solidus` we'll have `Solidus::SolidusShipwireConfig`

kennyadsl commented on lib/spree/solidus_shipwire_configuration.rb:
```diff
@@ -0,0 +1,22 @@
+    preference :logger,       :string,  default: false
+    preference :secret,       :string,  default: nil
+
+    def set_shipwire
```
maybe `setup_shipwire` sounds better.

kennyadsl commented on spec/controllers/shipwire_webhook_spec.rb:
```diff
@@ -0,0 +1,33 @@
+RSpec.describe Spree::ShipwireWebhookController, type: :controller do
+  controller Spree::ShipwireWebhookController do
+    def create
```
should this be `subscribe`?

mtylty replied:
I think the point here is testing exactly an action not called `subscribe`.

Anyway I think `subscribe` specs are missing?

DanielePalombo replied:
added in https://github.com/nebulab/solidus_shipwire/blob/stock-sync/spec/requests/spree/shipwire_webhook_spec.rb#L1

DanielePalombo committed:
Refactoring *SolidusShipwireConfig*

Refactoring from *SolidusShipwireConfig* to *ShipwireConfig*

DanielePalombo committed:
Specs for subscribe

Now `subscribe` return :not_found if the same path in `post` don't
exists in routes

DanielePalombo committed:
Change digest for shipwire signature

DanielePalombo committed:
solved signature problem

DanielePalombo commented:
Can I merge this PR?
